### PR TITLE
Update Identity positive test and add a negative test

### PR DESCRIPTION
**Pull request by hongdai**
* Update identity positve test per cert generation changes in bridge.
* Enable the positve test because it's now pass as both product and cert
functionalities are in place.
* Add a negative test.

Fixes issue #368

### CODE REVIEW

**Comment by hongdai** [comment]
@dotnet-bot test outerloop please

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
+using System.Collections;
 using System.ServiceModel;
+using System.ServiceModel.Security;
 using System.Text;
 using Xunit;
 
-namespace Security.TransportSecurity.Tests.Tcp
+using Infrastructure.Common;
+
+namespace Tcp_IdentityTests
```
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs]: I will remove the namespace as it will mess up the test result report.

**Comment by hongdai** [comment]
@dotnet-bot test outerloop please

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs [hunk 2]
```diff
@@ -83,16 +84,51 @@ public static string GetResourceAddress(string resourceName)
...
+                Dictionary<string, string> resources = null;
+                if (s_BridgeStatus == BridgeState.Started)
                 {
-                    resourceAddress = MakeEndpointResourcePutRequest(resourceName);
-                    _Resources.Add(resourceName, resourceAddress);
+                    if (!_Resources.TryGetValue(resourceName, out resources))
+                    {
+                        resources = MakeEndpointResourcePutRequest(resourceName);
+                        resources.TryGetValue(EndpointResourceResponseUriKeyName, out resourceAddress);
+                        _Resources.Add(resourceName, resources);
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs]: Nit: do this resource.TryGetValue just in one place, after the 'if' block.  Don't need the 'else' block.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 3]
```diff
@@ -176,10 +176,13 @@ public static void InstallLocalCertificateFromBridge()
                 }
             }
 
-            InstallCertificateToMyStore(certificateToInstall);
-
-            // We also need to install the root cert if we install a local cert
-            InstallRootCertificateFromBridge();
+            // certificateToInstall could be null if local cert has been installed in the previous tests
+            if (certificateToInstall != null)
+            {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: This comment and test makes me nervous.  Tests can run in parallel.  Can more than one set this value?  What happens if 2 tests run concurrently and change this value?  Can one test be in the middle of this != null test and have another test change it?  If so, we need to make this thread safe with a lock, or we need tests to pass in this information as they need it.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Security.TransportSecurity.Tests.csproj [hunk 4]
```diff
@@ -39,6 +39,10 @@
             <Project>{9098B41C-9C2E-4FC6-B7D8-FC3411736A22}</Project>
             <Name>ScenarioTests.Common</Name>
         </ProjectReference>
+    	<ProjectReference Include="$(WcfInfrastructureCommonProj)">
+      	    <Project>{AFD69665-89A3-42AE-A32E-AB2CBBE6EE7E}</Project>
+            <Name>Infrastructure.Common</Name>
+        </ProjectReference>
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Security.TransportSecurity.Tests.csproj]: This new project reference is a hint the ToF wrapper project living in TFS will also need to be updated to do the same thing.  That would require a TFS checkin separate from this one, and it could probably be done first to without harm.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs [hunk 5]
```diff
@@ -1,43 +1,75 @@
...
+        IWcfService serviceProxy = factory.CreateChannel();
+
+        try
         {
-            string testString = "Hello";
+            var result = serviceProxy.Echo(testString);
+            Assert.Equal(testString, result);
+        }
+        finally
+        {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs]: Nit: should probably use the close pattern Stephen has been using (do the close before the finally, and then have the finally call the testhelper to close any remaining CommunicationObjects

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs (hunk 3 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: certificateToInstall  is a local variable, not shared. Thus no other test can change it.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Security.TransportSecurity.Tests.csproj (hunk 4 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Security.TransportSecurity.Tests.csproj]: Will do.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs (hunk 3 above)
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: So maybe we should clarify the comment to say that some tests pass in null?  The phrase "previous test" makes this sound like it might be affected by test order.  NIT only.

Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: I will clarify it.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs (hunk 2 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs]: Will change.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs (hunk 5 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs]: Will do.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 6]
```diff
@@ -176,10 +176,13 @@ public static void InstallLocalCertificateFromBridge()
                 }
             }
 
-            InstallCertificateToMyStore(certificateToInstall);
-
-            // We also need to install the root cert if we install a local cert
-            InstallRootCertificateFromBridge();
+            // certificateToInstall could be null in the case the user certification exists
+            if (certificateToInstall != null)
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: I'd rather you check for `if (!foundUserCertificate)` here.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 7]
```diff
@@ -176,10 +176,13 @@ public static void InstallLocalCertificateFromBridge()
                 }
             }
 
-            InstallCertificateToMyStore(certificateToInstall);
-
-            // We also need to install the root cert if we install a local cert
-            InstallRootCertificateFromBridge();
+            // certificateToInstall could be null in the case the user certification exists
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: NIT: I don't quite think this is accurate. If !foundUserCertificate, certificateToInstall will not be null. The reason why these calls are here is because the Install*Certificate calls also take the same lock (s_certificateLock)

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs [hunk 8]
```diff
@@ -1,43 +1,77 @@
...
-            binding.Security.Mode = SecurityMode.Transport;
-            binding.Security.Transport.ClientCredentialType = TcpClientCredentialType.None;
+            factory.Close();
+        }
+        finally
+        {
+            if (factory != null && factory.State != CommunicationState.Closed)
+            {
+                factory.Abort();
+            }
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs]: The new pattern we were using is: 

        finally 
        {
            ScenarioTestHelpers.CloseCommunicationObjects(factory); 
        }

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs [hunk 9]
```diff
@@ -13,10 +13,11 @@ namespace Infrastructure.Common
     public static class BridgeClient 
     {
         private const string EndpointResourceResponseUriKeyName = "uri";
+        private const string EndpointResourceResponseFQHNKeyName = "hostname";
         private const string EndpointResourceRequestNameKeyName = "name";
 
         private static object s_thisLock = new object();
-        private static Dictionary<string, string> _Resources = new Dictionary<string, string>();
+        private static Dictionary<string, Dictionary<string, string>> _Resources = new Dictionary<string, Dictionary<string, string>>();
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs]: Please comment here on what the key and value are here

File: src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs [hunk 10]
```diff
@@ -145,6 +145,14 @@ public static string Tcp_VerifyDNS_Address
...
     }
 
+    public static string Tcp_VerifyDNS_HostName
+    {
+        get
+        {
+            return BridgeClient.GetResourceHostName("WcfService.TestResources.TcpVerifyDNSResource");
+        }
+    }
+
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs]: This seems like a weird of the Endpoints.cs class. Not blocking though, let's just get this in and we can talk about refactoring the BridgeClient classes

**Comment by iamjasonp** [comment]
I don't have many comments other than the NIT level comments. LGTM once you're done with those; please don't forget the TFS checkin.

File: src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs (hunk 10 above)
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs]: Agree to discuss naming.  Endpoints offer just offer services or certificates.  They don't "verify". NIT

**Comment by roncain** [comment]
LGTM, subject to the NIT level issues already mentioned

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs (hunk 7 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: certificateToInstall is null when foundUserCertificate is true, as the isLocalKeyName is not set, thus it continues to  InstallCertificateToMyStore.

The bug here might be why isLocalKeyName is not set from server. I'll open a issue to investigate it. Don't want to block this checkin.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs (hunk 6 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: See blow.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs (hunk 8 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/IdentityTests.cs]: I'll use the new pattern.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs (hunk 9 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClient.cs]: Will do.

File: src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs (hunk 10 above)
Comment (hongdai) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs]: For this particular endpoint, it's designed for only TCP Dns identity tests. I'm open to better naming.

**Comment by hongdai** [comment]
@dotnet-bot test outerloop please
